migrate ors to java 17

MichaelsJP opened the pull request:
Rewrite some old Java 11 API calls in the ors backend code, migrating them to Java 17 standards.

### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box
when you have checked you have done them): -->
- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.

Closes #1347 

### Information about the changes
- Key functionality added: Migration from Java 11 to 17.
- Reason for change: Move to a faster and more recent Java version. Additionally, the first end-of-life for Java 11 was already reached for OpenJDK. Others will follow in the subsequent years.

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
-

### Required changes to ors config (if applicable)
-

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

jlink commented:
I think that changing code to new Java syntax that’s also under heavy change in other PRs (eg api tests) is a bad idea. We shouldn’t have so many long lasting PRs that overlap. The later merger will have to do massive rework.

jlink reviewed (commented):
IMO this PR is MUCH too big. Everything except what's necessary to make the code compile and run with Java 17 is unnecessary for several reasons:
- Changing 80+ files in a long running PR will make merging later very involved
- Changing to a newer language syntax as such is no value in itself, especially that it always brings the danger of breaking code by mistake
- Which parts of the new language features should be rolled-out to all code is IMO something the team should decide
- Most of those changes can be done in small chunks, which is much less risky and easier to review

Therefore my recommendation is to strip this PR down to the bare necessities.

MichaelsJP committed:
chore(java): upgrade backend and api-tests to java 17

MichaelsJP committed:
build(workflow): Github workflow runner version to java 17

The workflow java version needs to be raised to 17 in order to run the migrated code base.

MichaelsJP committed:
chore(docker): Adjust the Dockerfile to java 17

The newest java 17 + alpine build ships with a well configured musl and brings the musl locale packages. There is no need anymore to build glibc for locales or c library access.

MichaelsJP commented:
@jlink Absolutely right. I saved most of the changes for a later cleanup and only left the necessary changes in.

jlink reviewed (approved):
LGTM

MichaelsJP committed:
docs(documentation): Adjust the documentation to java 17

MichaelsJP committed:
docs(changelog): Add the java 17 migration to the changelog
